opentelemetry-api/src/opentelemetry/context/__init__.py

```diff
@@ -104,16 +86,29 @@ def get_current() -> Context:
     return _RUNTIME_CONTEXT.get_current()  # type:ignore
 
 
-def set_current(context: Context) -> Context:
-    """To associate a context with program execution, the Context
-    API provides a function which takes a Context.
+def attach(context: Context) -> object:
+    """Associates a Context with the caller's current execution unit. Returns
+    a token that can be used to restore the previous Context.
+
+    Args:
+        context: The Context to set as current.
+    """
+    get_current()
+
+    return _RUNTIME_CONTEXT.set_current(context)  # type:ignore
+
+
+def detach(token: object) -> None:
+    """Resets the Context associated with the caller's current execution unit
+    to the value it had before attaching a specified Context.
 
     Args:
-        context: The context to use as current.
+        token: The Token that was returned by a previous call to attach a Context.
     """
-    old_context = get_current()
-    _RUNTIME_CONTEXT.set_current(context)  # type:ignore
-    return old_context
+    try:
+        _RUNTIME_CONTEXT.reset(token)  # type: ignore
+    except (AttributeError, TypeError, ValueError):
```

Comment:
Why not to capture all exceptions?

Reply:
Mostly a preference here, i could change it to all exceptions if you feel strongly about it.

Reply:
I feel like the catching specific exceptions may introduce more confusion than the value it provides.

I'm not sure what kind of exceptions resetting a token could cause, but it's entirely possible it results in a custom exception raised that is not caught. Then the behavior diverges regarding failure of the detach method from logging an error to raising an exception.

I'd say, as a user, I'd like to see the error handling behavior consistent: either catch all exceptions so I can assume that an error will be logged here in a failure scenario, or raise an exception so I can learn to capture exceptions around this method.

Reply:
I guess the spec calls from handling this defensively and never impacting the user app, so in that vein I"d catch and log all errors.

Reply:
updated to catch all